File: swift/RobotName.swift
import Foundation

class Robot {
    private var _name: String = ""
    private var _names: Set<String> = []
    
    var name: String {
        get {
            if _name.isEmpty {
                let letters = Array("ABCDEFGHIJKLMNOPQRSTUVWXYZ")
                
                repeat {
                    let index = Int(arc4random_uniform(24))
                    let number = arc4random_uniform(899) + 100
                    
                    _name = "\(letters[index])\(letters[index + 1])\(number)"
                } while _names.contains(_name)
            }
            
            return _name
        }
    }
    
    func resetName() {
        _name = ""
    }
}

File: clojure/robot_name.clj
(ns robot-name)

(defn generate-name
  "Generates a unique name for the robot."
  []
  (let [letters (apply str (repeatedly 2 #(char (+ (rand-int 26) 65))))
        number (rand-int 1000)]
    (format "%s%03d" letters number)))

(defn robot
  "Creates a robot, and names it."
  []
  (atom {:original-name (generate-name)}))

(defn robot-name
  "Returns the name of the robot."
  [r]
  (:original-name @r))

(defn reset-name
  "Resets the name of the robot."
  [r]
  (swap! r assoc :original-name (generate-name)))
